### ISSUE TITLE
Add Accessibility/ARIA compliance to all components

### ISSUE THREAD

**Issue by icfantv** [angular-ui/bootstrap#4772]
We need to review all the components for ARIA accessibility compliance and implement as appropriate per the [ARIA spec](http://www.w3.org/WAI/intro/aria).

Things we need to consider and evaluate:
- [Keyboard Navigation for Widgets](http://www.w3.org/TR/wai-aria-practices/#kbd_generalnav)
- [Keyboard Focus](http://www.w3.org/TR/wai-aria-practices/#kbd_focus)
- [ARIA States and Properties](http://www.w3.org/TR/wai-aria-practices/#aria-write)
- [Open Ajax Accessibility examples](http://oaa-accessibility.org/examples/)

Supported Components:

- [ ] Accordion ([docs](http://www.w3.org/TR/wai-aria-practices/#accordion))
- [ ] Alert ([docs](http://www.w3.org/TR/wai-aria-practices/#alert), [as modal](http://www.w3.org/TR/wai-aria-practices/#alertdialog))
- [ ] Buttons ([docs](http://www.w3.org/TR/wai-aria-practices/#button))
- [ ] Carousel ([docs](http://www.w3.org/WAI/tutorials/carousels/structure), [example](http://accessibility.athena-ict.com/aria/examples/carousel.shtml))
- [ ] Collapse ([docs](http://www.w3.org/TR/wai-aria-practices/#popupmenu))
- [ ] Datepicker ([docs](http://www.w3.org/TR/wai-aria-practices/#datepicker))
- [ ] Dropdown ([docs](http://www.w3.org/TR/wai-aria-practices/#popupmenu))
- [ ] Modal ([docs](http://www.w3.org/TR/wai-aria-practices/#dialog_modal))
- [ ] Pagination ([docs])
- [ ] Popover ([docs](http://www.w3.org/TR/wai-aria-practices/#tooltip))
- [ ] Progressbar ([docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_progressbar_role))
- [ ] Rating ([docs])
- [ ] Tabs ([docs](http://www.w3.org/TR/wai-aria-practices/#tabpanel))
- [ ] Timepicker ([docs](http://www.w3.org/TR/wai-aria-practices/#datepicker))
- [ ] Tooltip ([docs](http://www.w3.org/TR/wai-aria-practices/#tooltip))
- [ ] Typeahead ([docs](http://www.w3.org/TR/wai-aria-practices/#autocomplete))

**Comment by icfantv** [comment]
There's a good discussion and links in #3526.

**Comment by vcarreno96** [comment]
Is it possible to trace the closed sub issue for #3474 to a specific commit/release? I'm currently on version 0.12.1 of ui-bootstrap and need this particular accessibility fix. I would like to avoid going to a minor release if possible or else I would need a change log with breaking api changes. Your help is much appreciated!

**Comment by pkozlowski-opensource** [comment]
@vcarreno96 you can find all the details is https://github.com/angular-ui/bootstrap/blob/master/CHANGELOG.md

**Comment by NickHeiner** [comment]
I'm having a bit of a hard time figuring out how much ARIA work was actually done. None of the checkboxes in the original issue are checked. Some of the linked issues have landed as PRs, whereas others appear to have been closed as duplicates of this one. 

The reason I ask is that I'm working with a project on version `0.14.3`, and I'm trying to determine if upgrading to the latest version will yield accessibility improvements to the modal.

**Comment by icfantv** [comment]
@NickHeiner, IIRC, we've not done much and the release we targeted for ARIA support got hijacked and we never circled back.  It needs to be done so if you feel like submitting some quality PRs, please go ahead.

**Comment by NickHeiner** [comment]
Here's a PR! You can decide if it's quality 😸.

https://github.com/angular-ui/bootstrap/pull/6166

**Comment by jcwarfield** [comment]
Typeahead with custom HTML popup template specified - screen readers are not picking up the results displayed in the popup template. Seems the correct 'aria' tags are being generated but reader is silent, even when an item is 'active'.

**Comment by NickHeiner** [comment]
@jcwarfield can you try my PR (#6203) and tell me if that fixes your problem?

**Comment by jcwarfield** [comment]
Hi Nick, thanks for the quick response. I think your PR is a good addition, but does not fix the issue I'm having. Our uib-typeahead is using a custom popup template property (typeahead-popup-template-url), that points to an embedded template in the same html file (script type="text/ng-template" id="sectionList.html"). I think the problem may be that the custom template is being inserted into the DOM after the page is rendered, but before typing occurs, so the screen reader does not pick up the results in the popup. Is the template being injected after the page is rendered and a user begins typing?

**Comment by NickHeiner** [comment]
Ah yes, then my PR is not going to help you. That's a separate issue. And yes, I have seen it make a difference whether a DOM attribute exists when the elements are put into the DOM, as opposed to being added later.

**Comment by andypotts** [comment]
@jcwarfield, did you have any luck resolving the typeahead accessibility (screenreader) issue?

**Comment by meghanking** [comment]
Is there any update for these changes?

**Comment by jcwarfield** [comment]
Actually it's been over a year since working on that project, so it was still unresolved at the time I left.

**Comment by jrencz** [comment]
I'd like to mention that Modal component may not only take `modal` role. It may also be `alert` or `alertdialog`.

I considered implementing it over angular-ui-bootstrap in my application and I came into a conclusion that if backdrop is static (which in my case also means there's no "X" in the corner, but it's my app specific) then it's `alertdialog` (in my case it's always some way to interact).

I think there should be an optional param in `$uibModal.open` which says what kind of role should be applied. Now I would have to implement that with raw DOM api
